pkg/codegen/json/gen_program.go

```diff
@@ -0,0 +1,408 @@
+	variableJSON := make(map[string]interface{})
+	variableJSON["type"] = "LocalVariable"
+	variableJSON["name"] = variable.Name()
+	variableJSON["logicalName"] = variable.LogicalName()
+	variableJSON["value"] = transformExpression(variable.Definition.Value)
+	return variableJSON
+}
+
+func transformOutput(output *pcl.OutputVariable) map[string]interface{} {
+	outputJSON := make(map[string]interface{})
+	outputJSON["type"] = "OutputVariable"
+	outputJSON["name"] = output.Name()
+	outputJSON["logicalName"] = output.LogicalName()
+	outputJSON["value"] = transformExpression(output.Value)
+	return outputJSON
+}
+
+func transformConfigVariable(variable *pcl.ConfigVariable) map[string]interface{} {
+	variableJSON := make(map[string]interface{})
+	variableJSON["type"] = "ConfigVariable"
+	variableJSON["defaultValue"] = transformExpression(variable.DefaultValue)
+	variableJSON["name"] = variable.Name()
+	variableJSON["logicalName"] = variable.LogicalName()
+	switch variable.Type() {
+	case model.StringType:
+		variableJSON["configType"] = "string"
+	case model.NumberType:
+		variableJSON["configType"] = "number"
+	case model.IntType:
+		variableJSON["configType"] = "int"
+	case model.BoolType:
+		variableJSON["configType"] = "bool"
+	default:
+		variableJSON["configType"] = "unknown"
+	}
+
+	return variableJSON
+}
+
+func transformProgram(program *pcl.Program) map[string]interface{} {
+	programJSON := make(map[string]interface{})
+	nodes := make([]interface{}, 0, len(program.Nodes))
+	plugins := make([]interface{}, 0, len(program.Packages()))
+	for _, node := range program.Nodes {
+		switch node := node.(type) {
+		case *pcl.Resource:
+			transformedResource := transformResource(node)
+			nodes = append(nodes, transformedResource)
+		case *pcl.OutputVariable:
+			transformedOutput := transformOutput(node)
+			nodes = append(nodes, transformedOutput)
+		case *pcl.LocalVariable:
+			tranformedVariable := transformLocalVariable(node)
+			nodes = append(nodes, tranformedVariable)
+		case *pcl.ConfigVariable:
+			tranformedVariable := transformConfigVariable(node)
+			nodes = append(nodes, tranformedVariable)
+		}
+	}
+
+	for _, pkg := range program.Packages() {
+		pluginRef := map[string]interface{}{
+			"name":    pkg.Name,
+			"version": pkg.Version.String(),
+		}
+
+		plugins = append(plugins, pluginRef)
+	}
+
+	programJSON["nodes"] = nodes
+	programJSON["plugins"] = plugins
+	return programJSON
+}
+
+//go:embed pcl-json-schema.json
+var pclJSONSchema string
+
+func GenerateProgram(program *pcl.Program) (map[string][]byte, hcl.Diagnostics, error) {
+	files := make(map[string][]byte)
+	diagnostics := hcl.Diagnostics{}
+	programJSON := transformProgram(program)
+	schemaCompiler := jsonschema.NewCompiler()
+	schemaCompiler.LoadURL = func(_ string) (io.ReadCloser, error) {
+		return io.NopCloser(strings.NewReader(pclJSONSchema)), nil
+	}
+
+	schema := schemaCompiler.MustCompile("blob://irrelevant-url")
```

Comment:
We normally give the github url here: e.g. in core.go we use "https://github.com/pulumi/pulumi/blob/master/sdk/go/common/apitype/deployments.json".

Arguably these ought to point to the raw version of the urls 🤔  but we using LoadURL so don't actually hit them right now.

Or "blob://<filename>.json" so here you'd have "blob://pcl-json-schema.json".

Reply:
Alright, let me fix that! Will use `blob://pcl-json-schema.json` for this.